MonkeyWrench.Web.WebService/Scheduler/Scheduler.cs

```diff
@@ -400,11 +395,43 @@ private static void AddWork (DB db, List<DBHost> hosts, List<DBLane> lanes, List
 								}
 							}
 
-							sql.AppendFormat ("UPDATE RevisionWork SET state = {0} WHERE id = {1} AND state = 10;", (int) (has_dependencies ? DBState.DependencyNotFulfilled : DBState.NotDone), revisionwork.id);
+							sql.AppendFormat ("UPDATE RevisionWork SET state = 11 WHERE id = {0} AND state = 10;", revisionwork.id);
+							if (!editedHostLanes.Contains (hostlane)) editedHostLanes.Add (hostlane);
+						}
+					}
+				}
 
+				foreach (var editedHostLane in editedHostLanes) {
+					int mostRecent = -1;
+					int mostRecentState = -1;
+					using (IDbCommand cmd = db.CreateCommand ()) {
+						cmd.CommandText = @"
+SELECT RevisionWork.id, RevisionWork.state
+FROM RevisionWork
+INNER JOIN Revision ON Revision.id = RevisionWork.revision_id
+WHERE
+     RevisionWork.host_id = @host_id
+ AND RevisionWork.lane_id = @lane_id
+ORDER BY Revision.Date DESC
+LIMIT 1
+";
+						DB.CreateParameter (cmd, "host_id", editedHostLane.host_id);

```

Comment:
I think you can do this in one query for all RevisionWorks:

```sql
SELECT RevisionWork.id, MAX (Revision.date)
FROM RevisionWork
INNER JOIN Revision ON Revision.id = RevisionWork.revision_id
GROUP BY RevisionWork.id
```

and then iterate over the returned id, replacing it in the following query:

```sql
UPDATE RevisionWork SET state = {new_state} WHERE id = {id} AND state = 10;

UPDATE Work
SET state = {new_state}
INNER JOIN RevisionWork ON Work.revisionwork_id = RevisionWork.id
WHERE revisionwork_id = {id} and RevisionWork.state = 10;
```